File: examples/t01_build_your_first_tree.cpp
#include "behaviortree_cpp/bt_factory.h"

//#define MANUAL_STATIC_LINKING

#ifdef MANUAL_STATIC_LINKING
#include "dummy_nodes.h"
#endif

using namespace BT;

/** Behavior Tree are used to create a logic to decide what
 * to "do" and when. For this reason, our main building blocks are
 * Actions and Conditions.
 *
 * In this tutorial, we will learn how to create custom ActionNodes.
 * It is important to remember that NodeTree are just a way to
 * invoke callbacks (called tick() ). These callbacks are implemented by the user.
 */

// clang-format off
static const char* xml_text = R"(

 <root BTCPP_format="4" >

     <BehaviorTree ID="MainTree">
        <Sequence name="root_sequence">
            <CheckBattery   name="battery_ok"/>
            <OpenGripper    name="open_gripper"/>
            <ApproachObject name="approach_object"/>
            <CloseGripper   name="close_gripper"/>
        </Sequence>
     </BehaviorTree>

 </root>
 )";

// clang-format on

int main()
{
  // We use the BehaviorTreeFactory to register our custom nodes
  BehaviorTreeFactory factory;

  /* There are two ways to register nodes:
    *    - statically, i.e. registering all the nodes one by one.
    *    - dynamically, loading the TreeNodes from a shared library (plugin).
    * */

#ifdef MANUAL_STATIC_LINKING
  // Note: the name used to register should be the same used in the XML.
  // Note that the same operations could be done using DummyNodes::RegisterNodes(factory)

  using namespace DummyNodes;

  // The recommended way to create a Node is through inheritance.
  // Even if it requires more boilerplate, it allows you to use more functionalities
  // like ports (we will discuss this in future tutorials).
  factory.registerNodeType<ApproachObject>("ApproachObject");

  // Registering a SimpleActionNode using a function pointer.
  // you may also use C++11 lambdas instead of std::bind
  factory.registerSimpleCondition("CheckBattery",
                                  [&](TreeNode&) { return CheckBattery(); });

  //You can also create SimpleActionNodes using methods of a class
  GripperInterface gripper;
  factory.registerSimpleAction("OpenGripper", [&](TreeNode&) { return gripper.open(); });
  factory.registerSimpleAction("CloseGripper",
                               [&](TreeNode&) { return gripper.close(); });

#else
  // Load dynamically a plugin and register the TreeNodes it contains
  // it automated the registering step.
  factory.registerFromPlugin("../sample_nodes/bin/libdummy_nodes_dyn.so");
#endif

  // Trees are created at deployment-time (i.e. at run-time, but only once at the beginning).
  // The currently supported format is XML.
  // IMPORTANT: when the object "tree" goes out of scope, all the TreeNodes are destroyed
  auto tree = factory.createTreeFromText(xml_text);

  // To "execute" a Tree you need to "tick" it.
  // The tick is propagated to the children based on the logic of the tree.
  // In this case, the entire sequence is executed, because all the children
  // of the Sequence return SUCCESS.
  tree.tickWhileRunning();

  return 0;
}

/* Expected output:
*
       [ Battery: OK ]
       GripperInterface::open
       ApproachObject: approach_object
       GripperInterface::close
*/

File: python/examples/t01_build_your_first_tree.py
"""t01 — Build your first tree.

Shows the minimum needed to register a Python action, build a tree from
XML, and tick it to completion.

This tree runs two custom actions in sequence: a "check" that returns
SUCCESS or FAILURE based on a boolean, and a "say" that prints a line.

Run: python t01_build_your_first_tree.py
Expected output:
    [check] battery_ok=True
    [say] hello from pybt
    final status: SUCCESS
"""

from __future__ import annotations

import pybt

XML = """
<root BTCPP_format="4">
  <BehaviorTree ID="Main">
    <Sequence>
      <CheckBatteryOk/>
      <SaySomething/>
    </Sequence>
  </BehaviorTree>
</root>
"""


class CheckBatteryOk(pybt.SyncActionNode):
    def tick(self) -> pybt.NodeStatus:
        battery_ok = True
        print(f"[check] battery_ok={battery_ok}")
        return pybt.NodeStatus.SUCCESS if battery_ok else pybt.NodeStatus.FAILURE


class SaySomething(pybt.SyncActionNode):
    def tick(self) -> pybt.NodeStatus:
        print("[say] hello from pybt")
        return pybt.NodeStatus.SUCCESS


def main() -> int:
    factory = pybt.BehaviorTreeFactory()
    factory.register_node_type(CheckBatteryOk, "CheckBatteryOk")
    factory.register_node_type(SaySomething, "SaySomething")
    tree = factory.create_tree_from_text(XML)
    status = tree.tick_while_running()
    print(f"final status: {status.name}")
    return 0 if status == pybt.NodeStatus.SUCCESS else 1


if __name__ == "__main__":
    raise SystemExit(main())
